src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,272 @@
+
+                var modifiedRoot = documentEditor.GetChangedRoot();
+
+                // add an empty document to solution, so that we'll have options from the right context.
+                var solutionWithNewDocument = projectToBeUpdated.Solution.AddDocument(newDocumentId, newDocumentName, string.Empty/*, folders, fullFilePath*/);
+
+                // update the text for the new document
+                solutionWithNewDocument = solutionWithNewDocument.WithDocumentSyntaxRoot(newDocumentId, modifiedRoot, PreservationMode.PreserveIdentity);
+
+                // get the updated document, perform clean up like remove unused usings.
+                var newDocument = solutionWithNewDocument.GetDocument(newDocumentId);
+                return await CleanUpDocumentAsync(newDocument, cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<Solution> UpdateSourceDocumentAsync(
+                Document sourceDocument, TTypeDeclarationSyntax typeNode, CancellationToken cancellationToken)
+            {
+                var documentEditor = await DocumentEditor.CreateAsync(sourceDocument, cancellationToken).ConfigureAwait(false);
+
+                AddPartialModifiersToTypeChain(documentEditor, typeNode);
+
+                if (!_makeTypePartial)
+                {
+                    documentEditor.RemoveNode(typeNode, SyntaxRemoveOptions.KeepNoTrivia);
+                }
+
+                var updatedDocument = documentEditor.GetChangedDocument();
+                // TODO: make this optional -- shouldn't touch other parts of source document unless asked to. 
+                return await CleanUpDocumentAsync(updatedDocument, cancellationToken).ConfigureAwait(false);
+            }
+
+            private SyntaxNode GetMemberToAdd(SyntaxGenerator generator, string newDocumentName)
+            {
+                if (!_makeOuterTypesPartial)
+                {
+                    return null;
+                }
+
+                // we recognize filenames of that follow the pattern "OuterClass.InnerClass.cs" 
+                // and generate code for this file, which will look like so:
+                // a partial definition of OuterClass and a nested private InnerClass.
+                var nameParts = Path.GetFileNameWithoutExtension(newDocumentName).Split(new char[] { '.' });
+                if (nameParts == null || nameParts.Count() != 2)
+                {
+                    return null;
+                }
+
+                var innerTypeName = nameParts[1];
+                return generator.ClassDeclaration(innerTypeName, accessibility: Accessibility.Private);
+            }
+
+            // TODO: look to optimize the number of Ancestors and Descendants calls here.
+            private static IEnumerable<SyntaxNode> GetMembersToRemove(
+                SyntaxNode root, TTypeDeclarationSyntax typeNode, bool removeDescendentsOfSelf)
+            {
+                var ancestorsAndSelfToKeep = typeNode
```

Comment:
Oh... i see.  that's really slick!   I like that approach.  But yeah, i feel that this needs good comments, and can probably become a lot simpler.